### PR TITLE
SPI & entry clear in refresh_blackmap

### PR DESCRIPTION
**Pull request by beeender**
- refresh_blackmap doesn't need SPI calls.
- Clearing should happen within the same lock of refreshing entries.
  Since there will be a short time that other processes see a empty
  blackmap while checking the hard limit.

### CODE REVIEW

File: quotamodel.c [hunk 1]
```diff
@@ -1860,6 +1847,12 @@ refresh_blackmap(PG_FUNCTION_ARGS)
 
 	/* Flush the content of local_blackmap to the global blackmap. */
 	LWLockAcquire(diskquota_locks.black_map_lock, LW_EXCLUSIVE);
+
+	/* Clear blackmap entries. */
+	hash_seq_init(&hash_seq, disk_quota_black_map);
+	while ((blackmapentry = hash_seq_search(&hash_seq)) != NULL)
+		hash_search(disk_quota_black_map, &blackmapentry->keyitem, HASH_REMOVE, NULL);
+
 	hash_seq_init(&hash_seq, local_blackmap);
```
Comment (Sasasu) [inline quotamodel.c:1856]: https://github.com/greenplum-db/diskquota/pull/138#discussion_r807504101

local_blackmap can be a simple sorted array and not a global variable. I want to put this thing aside to do the release.
but we still fixing some rare bugs, this is not the right way.

this PR is OK to me, but this part code needs a big but simple refactor after we push a release.

**Review by Sasasu — APPROVED** [review]

File: quotamodel.c (hunk 1 above)
Comment (beeender) [reply quotamodel.c:1856]: > but we still fixing some rare bugs, this is not the right way.

yeah, but this one costs no effort. Just saw it and fix it.

**Review by higuoxing — APPROVED** [review]
Good catch! LGTM with one nit.

**Comment by beeender** [comment]
CI fails, not sure if it is related. Will check before merge

**Review by Sasasu — APPROVED** [review]

**Comment by Sasasu** [comment]
I made a branch to resolve the conflicts. https://github.com/greenplum-db/diskquota/commit/812747ec8d01d107b3f02cbba75d15eb834bfbe3

**Pull request merged**